D2k: Add map importer

penev92 opened the pull request:
This is my initial attempt to create a map importer for Dune 2000 .MAP files. (#4281)
Current status:
- Imports terrain
- Imports resources (spice)
- Imports actors (neutral, atreides and harkonnen; I will add the rest soon)

Issues:
- Missing tiles from arrakis.yaml tileset file  (#5875)
- Currently cannot add player references for mission maps

This is 90% working with skirmish maps (from what I've seen), excluding 10% for the missing tiles.
While this will import mission maps as well, the player reference issue does cause some problems there. This can be manually fixed in the `map.yaml` for now. I plan on fixing it properly when there is a Dune 2000 mission file reader implemented (#4280).

penev92 committed:
Initial implementation of D2k legacy map importer

Improve error handling; fix utility command description

Revert output location to working directory for both map importers

Rework how the map tiles get populated

Remove actor comments

Mailaender commented:
You will need to rebase this and thanks to https://github.com/OpenRA/OpenRA/pull/7245 Travis will tell you the StyleCop violations.

penev92 commented:
OK. Will get back to this soon (vacation is nearing an end and the sandworms are stuck for the moment).

penev92 commented:
Rebased, squashed, StyleCop'd and added Corrino and Fremen actor codes.

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,357 @@
+			{ 634, Pair.New("ROCKETTOWERA", "Fremen") },          //// Fremen rocket turret
+			{ 635, Pair.New("RESEARCHA", "Fremen") },             //// Fremen Ix research center
+			////{ 636, Pair.New("REPAIRA", "Fremen") },// Fremen repair pad
+			////{ 637, Pair.New("", "Fremen") },// Fremen sietch
```
"SIETCH" should work. It is already set up. https://github.com/OpenRA/OpenRA/blob/bleed/mods/d2k/rules/structures.yaml#L705

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,357 @@
+			{ 628, Pair.New("PALACEC", "Fremen") },               //// Imperial palace under Fremen control
+			{ 629, Pair.New("SILOA", "Fremen") },                 //// Fremen silo
+			{ 630, Pair.New("HEAVYA", "Fremen") },                //// Fremen heavy factory
+			////{ 631, Pair.New("REPAIRA", "Fremen") },// Fremen repair pad
```
Not sure why this is commented out.

penev92 replied:
I wanted to somehow mark the fact that both this and 636 say "Fremen repair pad". Obviously one of them is wrong, but I haven't been able to test that yet.

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,357 @@
+			{ 594, Pair.New("ROCKETTOWERH", "Corrino") },         //// Emperor rocket turret
+			{ 595, Pair.New("RESEARCHH", "Corrino") },            //// Emperor Ix research center
+			{ 596, Pair.New("STARPORTC", "Corrino") },            //// Emperor starport
+			////{597, Pair.New("", "Corrino") },// Emperor sietch
```
Again, `SIETCH` should just work.

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,357 @@
+			{ 633, Pair.New("HIGHTECHA", "Fremen") },             //// Fremen high tech factory
+			{ 634, Pair.New("ROCKETTOWERA", "Fremen") },          //// Fremen rocket turret
+			{ 635, Pair.New("RESEARCHA", "Fremen") },             //// Fremen Ix research center
+			////{ 636, Pair.New("REPAIRA", "Fremen") },// Fremen repair pad
```
This would be the Fremen starport assuming this is symmetrical like all the Dune 2000 assets.

penev92 replied:
You are right. I'll use that one.

penev92 commented:
Added Smugglers and Mercenaries and suggested structures.

penev92 commented:
I may be biased, but I think this is ready to merge, at least feature-wise. The only thing left on that front is integrating with the MIS file reader (if anyone is interested in that, [here is its current output](https://www.dropbox.com/s/itprsvz803pgrzc/mission.txt)), but that's just sugarcoating.
Only thing left is to check the code style, so I hope we can get this is before the final release.

RoosterDragon commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,415 @@
+				}
+		}
+
+		private TerrainTile GetTile(int tileIndex)
```
You can elide 'private' here.

RoosterDragon commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,415 @@
+				throw new InvalidOperationException("Tile with index {0} could not be found in the tileset YAML file!".F(tileIndex));
+
+			var templateIndex = template.Id;
+			byte frameIndex = 0;
```
```var frameIndex = (byte)Array.IndexOf(template.Frames, tileIndex);```

penev92 replied:
Won't that default to -1?

RoosterDragon replied:
Yes, though if I understand the rest of the code correctly, you've already assured that tileIndex is in the array anyway.

penev92 replied:
Ah, looks like you're right.

RoosterDragon commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,415 @@
+						indexOnMap = 0;
+
+					map.MapTiles.Value[locationOnMap] = terrainTiles[indexOnMap];
+					
```
Lamest comment I'll probably ever make: There is trailing whitespace on this otherwise empty line.

Mailaender commented:
http://cloudstor.es/f/dKS/ has some maps for testing. However I only get

```
Unhandled Exception:
System.IndexOutOfRangeException: Array index is out of range.
  at OpenRA.Mods.D2k.UtilityCommands.ImportD2kMapCommand.Run (OpenRA.ModData modData, System.String[] args) [0x00017] in /OpenRA.Mods.D2k/UtilityCommands/ImportD2kMapCommand.cs:31 
  at OpenRA.Utility.Program.Main (System.String[] args) [0x0015a] in /OpenRA.Utility/Program.cs:54 
[ERROR] FATAL UNHANDLED EXCEPTION: System.IndexOutOfRangeException: Array index is out of range.
  at OpenRA.Mods.D2k.UtilityCommands.ImportD2kMapCommand.Run (OpenRA.ModData modData, System.String[] args) [0x00017] in /OpenRA.Mods.D2k/UtilityCommands/ImportD2kMapCommand.cs:31 
  at OpenRA.Utility.Program.Main (System.String[] args) [0x0015a] in /OpenRA.Utility/Program.cs:54
```

when trying to import those.

penev92 commented:
I guess you didn't provide a tileset name?
That exception is because of the command-line arguments you give the utility.
Try with
`d2k --import-d2k-map "<path>\4-Side.map" BLOXBASE`

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/ImportD2kMapCommand.cs:
```diff
@@ -0,0 +1,38 @@
+	{
+		public string Name { get { return "--import-d2k-map"; } }
+
+		[Desc("FILENAME", "Convert a legacy INI/MPR map to the OpenRA format.")]
```
This has to be
```csharp
[Desc("FILENAME", "TILESET", "Convert a legacy INI/MPR map to the OpenRA format.")]
```

then.

pchote replied:
The description shouldn't be referring to INI/MPR either.

Mailaender commented:
> I just ran into tile with ID 29 - crashed the importer regardless of the tileset, because the yaml file doesn't have it

I assume you mean frame 29 not ID 29, because that does exist in arrakis.yaml. Feel free to add the missing tiles.

Mailaender commented:
> Also expect crashes if either you enter the wrong tileset name (currently can only guess that by trial and error)

According to the documentation at http://d2kplus.com/wiki/index.php?title=Map_Files and the source code of http://d2kplus.com/maprenderer.php the tileset is not stored in the MAP file format and I don't know where else. You could however give nicer error message such as "wrong/missing tileset" instead of crashing with the IndexOutOfRangeException. If we glue this to http://resource.openra.net/upload/map/ a dropdown box can ensure the users only select valid tileset names.

penev92 commented:
Yes, frame 29.

The tileset is unfortunately stored in the MIS file, which looks like a big design fail to me because
1. The MIS file has nothing to do with the map (except that it contains the tileset name)
2. Only missions have a MIS file, which means you have to guess for regular maps

Also the IndexOufOfRangeException was about the command-line arguments. I do give nicer exceptions about missing tiles - [see here](https://github.com/OpenRA/OpenRA/pull/7135/files#diff-54b547985b15dbd63a5da7db1eb43a54R404). ;)

If we put a how-to-use description (in the wiki, for example), I'd suggest using either the [D2k+ map renderer](http://d2kplus.com/maprenderer.php) (I haven't tried it), the [D2k+ campaign map editor](http://d2kplus.com/campaignmap.php) (works quite nicely) or [my own map previewer](https://github.com/penev92/D2kMapPreviewer).
Actually, for best results I recommend using the last two in combination.
These will give you a very good idea what you're importing so you can compare the result with the original map and will also "tell" you which tileset to use when importing (most maps will have broken tiles (in the previews) if you use an incorrect tileset).

After importing you can use [my WIP oramap preview](https://github.com/penev92/OpenRA/tree/bleed_generateMapPreview) branch to generate a PNG of the generated `oramap` file and compare that to the PNG from the original `MAP` file.

penev92 committed:
Add missing tiles BLOXBGBS-29 (1x1) and BLOXBGBS-617 (2x2)

penev92 commented:
So I fixed the description, added that tile and made sure you don't get an invalid (but still present) `.oramap` file if the import crashes.

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/ImportD2kMapCommand.cs:
```diff
@@ -0,0 +1,42 @@
+
+		    var dir = Path.Combine(Environment.CurrentDirectory, "mods", "d2k", "maps", "Imported");
+		    if (!Directory.Exists(dir))
+		        Directory.CreateDirectory(dir);
```
Tabs instead of spaces and I believe this is not necessary as OpenRA will automatically create the path if it doesn't exist.

penev92 replied:
It doesn't.

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/ImportD2kMapCommand.cs:
```diff
@@ -0,0 +1,42 @@
+			if (map == null)
+				return;
+
+		    var dir = Path.Combine(Environment.CurrentDirectory, "mods", "d2k", "maps", "Imported");
```
This will crash if you don't run the utility with admin rights once installed. It should also not be hard-coded. Please change this to:

```csharp
var userMapFolder = Game.ModData.Manifest.MapFolders.First(f => f.Value == "User").Key;
	if (userMapFolder.StartsWith("~"))
		userMapFolder = userMapFolder.Substring(1);

var dir = Platform.UnresolvePath(userMapFolder);
```

Mailaender replied:
Also tabs instead of spaces please.

penev92 commented:
Updated btw.

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/ImportD2kMapCommand.cs:
```diff
@@ -0,0 +1,48 @@
+			if (userMapFolder.StartsWith("~"))
+				userMapFolder = userMapFolder.Substring(1);
+
+			var dir = Path.Combine(Platform.ResolvePath(userMapFolder), "Imported");
```
The ´/Imported/` folder placement is a bit annoying as you have to manually copy it to test it.

Mailaender commented:
:+1: Works well.

When looking at ugly hacks like

```bat
@c:
@cd c:\westwood\dune2000\data
@copy bloxice.b8 bloxbgbs.r8
@copy bloxice.b16 bloxbgbs.r16
@c:
@cd c:\
```

in `Tileset_Substitution.Zip` from http://cloudstor.es/f/dKS/ then it is clear how much the Dune 2000 community suffered from the redundant yet (somewhere else) hard-coded tileset settings.

penev92 commented:
Removed the `Imported` subfolder and the `Directory.Exists()` check.

pchote commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+			{ 194, Pair.New("SONICTANK", "Atreides") },
+
+			// Harkonnen:
+			{ 204, Pair.New("WALLH", "Harkonnen") },              //// Harkonnen wall
```
These comments don't tell us anything useful, and can be removed.

Mailaender replied:
Yes please. They are pretty redundant. It is not helpful if source code comments describe obvious things.

pchote commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+			map.RequiresMod = mod;
+			map.MakeDefaultPlayers();
+
+			// TODO: Address FixOpenAreas crash
```
This should probably be addressed.

penev92 replied:
I'm not sure. What is FixOpenAreas?
And why would the map importer need to randomize the terrain (if I remember correctly what I understood from it) ?

ScottNZ replied:
Notice that the terrain repeats on new blank maps in the editor. Fix Open Areas fixes that.

penev92 replied:
Yeah I understand it is good for randomizing when you are making a map from scratch, but do we want the map importer to do it?

Mailaender replied:
I think it is not needed for Dune 2000 imports which pick individual (pre-randomized) tile images:

![image](https://cloud.githubusercontent.com/assets/756669/6183003/3f562ec6-b34f-11e4-9ba8-a2a69780afb2.png)

This is only a problem in RA/TD when the map import marks terrain as Clear and everything is tile variant 0 which looks wallpaper like:

![image](https://cloud.githubusercontent.com/assets/756669/6183015/739d7036-b34f-11e4-8187-379989945f0f.png)

which gets converted by "Fix Open Areas" to

![image](https://cloud.githubusercontent.com/assets/756669/6183018/81c82be2-b34f-11e4-8bbd-8bb612eb8ce8.png)

I automatized this for both clear tile placement and on new map generation in both the legacy WinForms editor and the upcoming in-game one #4496 for better usability and because it was often forgot to manually hit that button.

x-a-n-a-x replied:
wow, nice work Mailaender !

Mailaender replied:
That was done a long time ago though. https://github.com/OpenRA/OpenRA/pull/4394

pchote commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+
+		void LoadMapInfo()
+		{
+			for (var index = 4; index < stream.Length; index += 4)
```
`while (stream.Position < stream.Length)` would be more to the point.

pchote commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+
+		void FillMap()
+		{
+			for (var y = 0; y < map.MapSize.Y; y++)
```
Style nit: The inner block has braces, so this outer block needs them too.

pchote commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+					// Hack for a "clear" tile
+					var indexOnMap = 0;
+
+					if (map.Contains(locationOnMap))
```
This is a strange way to parse things.  It would be simpler/clearer to loop scan through the tiles in the source file, and then calculate the equivalent OpenRA map position.

pchote commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+					// Actors
+					if (actorDataByActorCode.ContainsKey(tileSpecialInfo))
+					{
+						var kvp = actorDataByActorCode[tileSpecialInfo];
```
This will throw if the actor id isn't defined (like the 641 fremen).  I suggest using a `TryGetValue` and then log a message and continue if the actor isn't found.

penev92 replied:
That is exactly why I have `if (actorDataByActorCode.ContainsKey(tileSpecialInfo))` just above.

pchote replied:
Oh, I see.

pchote commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+						var a = new ActorReference(kvp.First)
+						{
+							new LocationInit(locationOnMap),
+							new OwnerInit(kvp.Second)
```
I don't see any code to set up the players defined in the actor lists, so this will generate an unusable map file(?).

penev92 replied:
Iirc it worked.
Also nobody is guaranteeing that you'd be able to take a Dune 2000 map, press a button and have a complete and playable OpenRA map.

penev92 replied:
Ah, my bad with the above comment, didn't look hard enough at what you were commenting on.
The `Actors` part will only be viable for a MAP file that comes together with a MIS file anyway (so only for missions - official or custom). And my plan was to make the utility read both of those simultaneously if they are available in order to populate the map.yaml with the appropriate player definitions.

Like I said earlier, right now it works for skirmish/multiplayer maps, but not for mission maps (in the Dune 2000 sense), where you have to open the map.yaml and fix it yourself.
This is just one of the reasons the MIS file reader is important to implement.

pchote commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+			}
+
+			if (template == null)
+				throw new InvalidOperationException("Tile with index {0} could not be found in the tileset YAML file!".F(tileIndex));
```
Logging a message and returning the default clear tile would be more modder-friendly.

penev92 replied:
Logging where?

pchote replied:
`System.Console`.

pchote commented:
See inline comments for a few code/style nits.  I haven't had a chance to test this yet.

Mailaender commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,419 @@
+			map.MakeDefaultPlayers();
+
+			// TODO: Address FixOpenAreas crash
+			// map.FixOpenAreas(rules);
```
Just remove it as I believe it is simply not needed for d2k here.

penev92 commented:
While this works now, @pchote's `d2k-raceart` branch is more than likely going to screw up the actor import because of the hardcoded actor type names.

Mailaender commented:
I would like to get this merged first and deal with the problem in the upcoming pull request.

penev92 commented:
Removed the actor type comments and rebased.

obrakmann commented on OpenRA.Mods.D2k/UtilityCommands/ImportD2kMapCommand.cs:
```diff
@@ -0,0 +1,45 @@
+			var dir = Platform.ResolvePath(userMapFolder);
+
+			var dest = Path.Combine(dir, map.Title + ".oramap");
+			map.Save(dest);
```
This crashes if the path doesn't exist yet.

penev92 replied:
How can you not have that directory?

obrakmann commented on OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs:
```diff
@@ -0,0 +1,405 @@
+
+				var locationOnMap = GetCurrentTilePositionOnMap();
+
+				Console.WriteLine(locationOnMap.X + "   " + locationOnMap.Y);
```
I'm sure this is left-over from debugging and will be removed before merging, right?

obrakmann commented:
Looks good to me overall. I tested it with a skirmish map from the archive Matt linked to above, which worked fine ingame without any changes. I can't vouch for the missions, though, as I don't have the original game.

Code looks good, too, modulo the two mentioned nits. :+1: Once those are fixed.

Also needs a rebase.

pchote commented on OpenRA.Mods.D2k/UtilityCommands/ImportD2kMapCommand.cs:
```diff
@@ -0,0 +1,45 @@
+			if (userMapFolder.StartsWith("~"))
+				userMapFolder = userMapFolder.Substring(1);
+
+			var dir = Platform.ResolvePath(userMapFolder);
```
This is inconsistent with the rest of our utility commands, and commandline application practices in general.  The map should be saved to the current working dir if an explicit output location isn't given.

penev92 commented:
Removed the debug output, rebased, squashed and reverted the output directory for both map importers to the working directory as requested by @pchote.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/bc3c2cd999260ee86af502c826e291daa8e7595a%5E...bc3c2cd999260ee86af502c826e291daa8e7595a)
